docs/guides/02_writing_commands.md

```diff
@@ -0,0 +1,233 @@
+# Writing commands
+
+## Intro
+
+Interacting with systems based with Core Library is based on a Command Query Responsibility Segregation ([CQRS](../basics/02_cqrs.md)) principle. As such, there are two main types of actions which can be performed in the system: command, and [query](./03_writing_queries.md) [^1]. In this guide we will focus on the former. A command is used the change the state of the system but it does not yield any results - it can be treated as write-only action.
+
+Each command consists of three parts:
+
+* Contract
+* Command validator
+* Command handler
+
+## Contract
+
+A contract, as the name suggests, defines the way a command can be invoked by the clients. Let's see how a contract for creating a project might look like.
+
+```csharp
+[AllowUnauthorized]
+public class CreateProject : ICommand
+{
+    public string Name { get; set; }
+
+    public static class ErrorCodes
+    {
+        public const int NameCannotBeEmpty = 1;
+        public const int NameTooLong = 2;
+    }
+}
+```
+
+The contract for a command is simply a class implementing the `ICommand` interface. The fields of a contract class define how the body of an HTTP request should look like. In this example, an example of a valid request body would be:
+
+```json
+{
+    "Name": "abc",
+}
+```
+
+Notice the `ErrorCodes` class. These define possible validation errors which can be expected by the clients. Each error code should be a unique (in a given contract) integer.
+
+Writing the contracts in that way makes it possible for web and mobile clients to generate their own version of contract classes with use of the Contract Generator which is also a part of CoreLibrary.
+
+Also, a contract can have authorization attributes which define rules which have to be satisfied for an actor to be able to invoke the command. To learn more about authorizers, see [authorization](./0X_authorization.md).
+
+## Command validator
+
+Command validator's main role is to ensure that the command invoked by the user is valid. The command validator should check the structure of the command itself, for example, check for required fields, check whether passed strings do not exceed maximum length etc. Also the validator should check the business logic of the command for example, when editing a project, a command validator should check whether or not the project exists, see [more complex commands](#more-complex-commands) for an example. Let's see how a validator for `CreateProject` command might look like:
```

Comment:
```suggestion
The command validator's main role is to ensure that the command invoked by the user is valid. The command validator should check the structure of the command itself, for example, check for required fields, check whether passed strings do not exceed maximum length etc. Also the validator should check the business requirements of the command. For example, when editing a project, it should check whether or not the project exists, see [more complex commands](#more-complex-commands) for an example. Let's see how a validator for `CreateProject` command might look like:
```

We should also add a single sentence of "why" we split validation and execution.

Also, something to add to backlog - the guidelines of writing business logic validators.